contracts/ShortNameClaims.sol

```diff
@@ -82,80 +106,193 @@ contract ShortNameClaims is Ownable {
+     * @param status The status to set - must be APPROVED or DECLINED.
+     */
+    function setClaimStatus(bytes32 claimId, Status status) public inPhase(Phase.REVIEW) {
+        // Only callable by owner or ratifier
+        require(msg.sender == owner() || msg.sender == ratifier);
+
+        // Can't set claim back to pending
+        require(status == Status.APPROVED || status == Status.DECLINED);
+
         Claim memory claim = claims[claimId];
         require(claim.paid > 0, "Claim not found");
 
-        claimCount--;
-        delete claims[claimId];
-        emit ClaimApproved(claimId);
+        if(claim.status == Status.PENDING) {
+          // Claim went from pending -> approved/declined; update counters
+          pendingClaims--;
+          unresolvedClaims++;
+        } else if(claim.status == Status.APPROVED) {
+          // Claim was previously approved; remove from approved map
+          approvedNames[claim.labelHash] = false;
+        }
 
-        registrar.register(uint256(claim.labelHash), claim.claimant, REGISTRATION_PERIOD);
-        address(uint160(registrar.owner())).transfer(claim.paid);
+        // Claim was just approved; check the name was not already used, and add
+        // to approved map
+        if(status == Status.APPROVED) {
+          require(!approvedNames[claim.labelHash]);
+          approvedNames[claim.labelHash] = true;
+        }
+
+        claims[claimId].status = status;
+        emit ClaimStatusChanged(claimId, status);
     }
 
-    function declineClaim(bytes32 claimId) public {
-        // Only callable by contract owner or claimant
-        require(msg.sender == owner() || msg.sender == claims[claimId].claimant);
+    /**
+     * @dev Sets the status of multiple claims. Callable only during the review
+     *      phase, and only by the owner or ratifier.
+     * @param approved A list of approved claim IDs.
+     * @param declined A list of declined claim IDs.
+     */
+    function setClaimStatuses(bytes32[] calldata approved, bytes32[] calldata declined) external {
+        for(uint i = 0; i < approved.length; i++) {
+            setClaimStatus(approved[i], Status.APPROVED);
+        }
+        for(uint i = 0; i < declined.length; i++) {
+            setClaimStatus(declined[i], Status.DECLINED);
+        }
+    }
 
+    /**
+     * @dev Resolves a claim. Callable by anyone, only in the final phase.
+     *      Resolving a claim either registers the name or refunds the claimant.
+     * @param claimId The claim ID to resolve.
+     */
+    function resolveClaim(bytes32 claimId) public inPhase(Phase.FINAL) {
         Claim memory claim = claims[claimId];
         require(claim.paid > 0, "Claim not found");
 
-        claimCount--;
+        if(claim.status == Status.APPROVED) {
+            registrar.register(uint256(claim.labelHash), claim.claimant, REGISTRATION_PERIOD);
+            address(uint160(registrar.owner())).transfer(claim.paid);
```

Comment:
`address(uint160(addr))` can we move this into a function that clearly names what it does. non experts need to google this every time when reading the code.